### ISSUE TITLE
Review logic for setting bucketname/endpoint in S3FileSystem

### ISSUE THREAD

**Issue by tony-johnson** [stefanofornari/aws-java-nio-spi-for-s3#4]
The existing code has:

        String host = uri.getHost(); int port = uri.getPort();
        if ((port > 0) || (host.indexOf(':') > 0)) {
            this.bucketName = uri.getPath().split(S3Path.PATH_SEPARATOR)[1];
            this.endpoint = host + ((port > 0) ? (":" + port) : "");
        } else {
            this.bucketName = host;
            this.endpoint = null;
        }
        
As far as I can see, host can never contain ":". In my case I switched it to:

        if ((port > 0) || (host.indexOf('.') > 0)) {
            this.bucketName = uri.getPath().split(S3Path.PATH_SEPARATOR)[1];
            this.endpoint = host + ((port > 0) ? (":" + port) : "");
        } else {
            this.bucketName = host;
            this.endpoint = null;
        }

(. instead of :) and this works if my host contains . even  if I do not explicitly specify a port. But perhaps it would be good to more clearly list what syntaxes are expected to be supported.

**Comment by stefanofornari** [comment]
You are right, this needs to be reviewed. After a thoughtfully discussion with the official library maintainer we reach to the conclusion that the best thing is to support 3rd party services with a dedicated schema, something like s3x:// where we can more freely manage the URL. The issue with AWS is that S3 URIs are defined and some aspect clash with a more natural (for not AWS S3 services) approach. 
In short: this will be addressed as well. Thanks for pointing it out.

**Comment by stefanofornari** [comment]
I have created a [new branch](https://github.com/stefanofornari/aws-java-nio-spi-for-s3/tree/s3x) with the new approach to support 3rd party services. As per the previous comment, I have created a new S3XFileSystemProvider that handles the URI schema s3x://. In this way the original S3FileSystemProvider manages S3 URIs, while the new one handles URI in the form:

`
s3x://[key:secret@]endpoint[:port]/bucket/key`

I therefore adjusted the logic to set bucket name and endpoint. My next step is to create a PR for the official library.

Feedback are welcome
:)

**Comment by stefanofornari** [comment]
The branch has been merged in awslabs and a new version should be pushed soon; I am going to close this issue as the new official version is released
